transformation/src/main/scala/org/broadinstitute/monster/hca/HcaUtils.scala

```diff
@@ -0,0 +1,119 @@
+    * @param context context of the main V2F pipeline
+    */
+  def getReadableFiles(jsonPath: String, context: ScioContext): SCollection[FileIO.ReadableFile] =
+    context.wrap {
+      context.pipeline.apply(
+        FileIO
+          .`match`()
+          .filepattern(jsonPath)
+          .withEmptyMatchTreatment(EmptyMatchTreatment.ALLOW_IF_WILDCARD)
+      )
+    }.applyTransform[ReadableFile](FileIO.readMatches())
+
+  /**
+    * Given the SCollection of ReadableFiles that contains JSONs convert each JSON to a Msg and get its filename.
+    *
+    * @param tableName the name of the JSON table that was converted to Msg
+    */
+  def jsonToFilenameAndMsg(tableName: String): SCollection[ReadableFile] => SCollection[(String, Msg)] =
+    _.transform(s"Extract $tableName JSON rows") {
+      _.map { file =>
+        (
+          file.getMetadata.resourceId.getFilename,
+          JsonParser.parseEncodedJson(file.readFullyAsUTF8String)
+        )
+      }
+    }
+
+  /**
+    *
+    * extract the necessary info from a metadata file and put it into a form that makes it easy to
+    * pass in to the table format
+    *
+    * @param entityType the metadata entity type to prepend for the id field
+    * @param fileName the raw filename of the metadata file
+    * @param metadata the content of the metadata file in Msg format
+    * @return a Msg object in the desired output format
+    */
+  def transformMetadata(entityType: String, fileName: String, metadata: Msg): Msg = {
+    // extract info from file name
+    val matches = HcaUtils.metadataPattern
+      .findFirstMatchIn(fileName)
+      .getOrElse(
+        throw new Exception(
+          s"transformMetadata: error when finding id and version from file named $fileName"
+        )
+      )
+    val entityId = matches.group(1)
+    val entityVersion = matches.group(2)
+    // and put in form we want
+    Obj(
+      mutable.LinkedHashMap[Msg, Msg](
+        Str(s"${entityType}_id") -> Str(entityId),
+        Str("version") -> Str(entityVersion),
+        Str("content") -> metadata
+      )
+    )
+  }
+
+  /**
+    * Read, transform, and write a given entity type.
+    *
+    * @param context context of the main pipeline
+    * @param inputPrefix the root directory containing JSONs to be converted
+    * @param outputPrefix the directory to write outputs to
+    * @param entityType the bucket name/table name to read/write from
+    */
+  def processMetadata(
+    context: ScioContext,
+    inputPrefix: String,
+    outputPrefix: String,
+    entityType: String
+  ): ClosedTap[String] = {
+    // get the readable files for the given input path
+    val readableFiles = getReadableFiles(
+      s"$inputPrefix/metadata/${entityType}/**.json",
+      context
+    )
+    // then convert json to msg and get the filename
+    val processedData = jsonToFilenameAndMsg(entityType)(readableFiles)
+      .withName(s"Convert ${entityType} from JSON to Msg in output form")
```

Comment:
Name here is pretty low-level, suggested alternative is:
```suggestion
      .withName(s"Pre-process ${entityType} metadata")
```